### PR TITLE
Add a lint script, eslint and OrbitDB linting rules

### PR DESCRIPTION
**Pull request by haadcode**
This PR will add linting to the project.

Depends on https://github.com/orbitdb/eslint-config-orbitdb/pull/6.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -19,8 +19,7 @@
   },
   "homepage": "https://github.com/orbitdb/orbit-db-identity-provider#readme",
   "devDependencies": {
-    "eslint": "^5.2.0",
-    "@orbitdb/eslint-config-orbitdb": "orbitdb/eslint-config-orbitdb#feat/self-standing",
+    "@orbitdb/lint": "haadcode/orbitdb-lint",
```
Comment (RichardLitt) [inline package.json]: Did we make a new one? :D

**Comment by RichardLitt** [comment]
@haadcode That PR was merged.

**Comment by haadcode** [comment]
After merging the dependant PR and having tried to setup linting again here, I believe we should go with another approach which I believe @thiagodelgado111 was going for initially.

First, I was wrong about the "dependencies" vs. "devDependencies" in the linked PR: they can't be de deps since dev deps are not transitive (as in, if eslint-config-orbitdb has dev deps, they won't get installed when eslint-config-orbitdb is used as a module, only when installed *inside* eslint-config orbitdb).

Second, given the way eslint works, it seems hard to set it up in a "light way" that we would only incude a config file as there are those deps needed for it to work.

So my proposal would be to create a separate module that pulls in all the needed dependencies, including eslint itself. Which I have done here https://github.com/haadcode/orbitdb-lint. I didn't want to put in under orbitdb org yet as to not create more confusion (which I suppose the commit above already did as @RichardLitt noticed it :)).

The general idea is that instead of developer installing linting with `npm i eslint-config-orbitdb` plus all the deps, the developer of a module would install `npm i @orbitdb/lint` package which would bring in "full linting solution" and the only thing the developer needs to do is to add .eslintrc and "lint" script to package.json. (as a side, I wonder if there's a way to automatically "setup" .eslintrc, can eslint take the config file from another directory? eg. node_modules/@orbitdb/lint/.eslintrc?)

That way it'd be fully stand-alone and no manual input is required from the developer to install additional packages. We would also be able to revert the PR I did to eslint-config-orbitdb and define the deps as "peerDependencies" as they originally were.

I feel this would be the cleanest solution while keeping the module responsibilities clear and provide a smooth dev experience.

What do you think?

**Comment by RichardLitt** [comment]
I'm confused. Why can't we use eslint-config-orbitdb as the package that draws in all of the deps, instead of making another package (orbitdb-lint)? Eslint should be a dependency of eslint-config-orbitdb, and any other needed modules for it to run. Then, when you install a module like orbit-db-identity-provider, it would install eslint-config-orbitdb as a devDep, and all deps of that module would also be installed. 

```
orbit-db-identity-provider
- dep: ...
- devDep: eslint-config-orbitdb
  - dep: eslint
  - devDep: [none]
```

The above seems to work more fluidly than 

```
orbit-db-identity-provider
- dep: ...
- devDep: orbitdb-lint
  - dep: eslit-config-orbitdb
    - dep: eslint (_currently defined as devDep; should probably be moved_)
    - devDep: none 
  - devDep: [none]
```

What am I missing? 

---

Regarding how to automatically set up an .eslintrc; I would suggest using the `package.json` field instead of an extra file (because it keeps the files down), and possibly setting up [a proper eslint config package or plugin](https://eslint.org/docs/user-guide/configuring#using-a-shareable-configuration-package). Otherwise, we can just add it to the [repo-template](https://github.com/orbitdb/repo-template).

**Comment by haadcode** [comment]
> Why can't we use eslint-config-orbitdb as the package that draws in all of the deps, instead of making another package (orbitdb-lint)? Eslint should be a dependency of eslint-config-orbitdb, and any other needed modules for it to run. 

I think the idea was (@thiagodelgado111 correct me if I'm wrong) the `eslint-config-orbitdb` would contain *only* the linting rules (=config), not "full linting solution".

With that I'm proposing above, the deps would look like:
```
orbit-db-identity-provider
- dep: ...
- devDep: orbitdb-lint
  - dep: eslint, eslit-config-orbitdb, <peer dependencies required by eslint-config-orbitdb>
  - devDep: [none]
```

**Comment by haadcode** [comment]
Ping @thiagodelgado111, any thoughts on the approach proposed?

**Comment by thiagodelgado111** [comment]
I don't know, I think it's unnecessary to add a dependency that adds eslint and add plugins and eslint to that package. 

I for one would simply go back to the previous setup, export linting rules and keep eslint/plugins as peer deps to signal they are needed... If the instructions on how to set it up are clear, it shouldn't be a problem.

**Comment by haadcode** [comment]
Given how the dependencies work in eslint (peerDeps) and its config, and while I agree it's not *necessary* (it would work as it is), providing a *better user experience* is absolutely something to aim for. Instead of *n* steps to install and configure the tools, it's *always* better if it can be done in one or two steps. It makes a huge difference: if there's "friction" to setting something up, especially when it comes to dev tools, the user quickly will think it's too much to read/learn/do and will switch to something else or come up with their own solution. The whole point of the proposed approach, along with the associated documentation changes earlier, is to make it as simple and smooth as possible for the user. I think that's desirable, always.

@thiagodelgado111 any reason you feel it's unnecessary?

**Comment by thiagodelgado111** [comment]
> The whole point of the proposed approach, along with the associated documentation changes earlier, is to make it as simple and smooth as possible for the user. I think that's desirable, always.

I agree with that!

> It makes a huge difference: if there's "friction" to setting something up, especially when it comes to dev tools, the user quickly will think it's too much to read/learn/do and will switch to something else or come up with their own solution. 

I don't necessarily agree with that, I think setting up eslint is something fairly simple nowadays. It requires just a bit of reading of eslint's docs. That's why, personally, I would go with the approach I think it's simpler, the one recommended by the eslint team. I'd never go against aiming for the best dev experience, maybe we should take it as a complaint to eslint instead

**Comment by RichardLitt** [comment]
How hard would it be to just throw up three quick packages that do all of these different iterations, and then write up the steps needed to use each one, so that we can compare them side by side in action? 

> It requires just a bit of reading of eslint's docs. 

Never underestimate how hard this is. ;) We're going to have to copy them to each repo, locally.

**Comment by RichardLitt** [comment]
I have a rather silly question that I can't seem to resolve. The only difference between OrbitDB style and Standard is ` 'brace-style': ['error', '1tbs', { allowSingleLine: false }]`. Is there a specific reason why we need this rule? We could just use standard and have a very easy install process.

**Comment by haadcode** [comment]
>  The only difference between OrbitDB style and Standard is 'brace-style': ['error', '1tbs', { allowSingleLine: false }]. Is there a specific reason why we need this rule? We could just use standard and have a very easy install process.

If that's the only difference, then let's go with standard and the easy install process 👍

**Comment by RichardLitt** [comment]
Note https://github.com/orbitdb/orbit-db-identity-provider/pull/14. I think this is the best way forward. Happy to go through and overhaul linting for all of the other repos.

**Comment by RichardLitt** [comment]
This was superseded by #14.

**Pull request closed without merging**